File: Go/ExampleTests/common/inputcontrolvalues_string.go
package common

type InputControlValuesString struct {
	Frame string
	Roll string
	Remaining string
}

func NewInputControlValuesStringFromSlice(v []string) InputControlValuesString {
	s := InputControlValuesString{}
	if len(v) > 0 { s.Frame = v[0] }
	if len(v) > 1 { s.Roll = v[1] }
	if len(v) > 2 { s.Remaining = v[2] }
	return s
}

// NewInputControlValuesStringFromText builds from the text form, e.g. Money as "25 USD".
func NewInputControlValuesStringFromText(text string) InputControlValuesString {
	parts := RequireTokens(text, 3, "InputControlValues")
	return InputControlValuesString{
		Frame: parts[0],
		Roll: parts[1],
		Remaining: parts[2],
	}
}

func (s InputControlValuesString) String() string {
	return Token(s.Frame) + " " + Token(s.Roll) + " " + Token(s.Remaining)
}

func (s InputControlValuesString) Equals(o InputControlValuesString) bool {
	return DNCEqual(s.Frame, o.Frame) &&
		DNCEqual(s.Roll, o.Roll) &&
		DNCEqual(s.Remaining, o.Remaining)
}

func EqualInputControlValuesStringSlices(a, b []InputControlValuesString) bool {
	if len(a) != len(b) { return false }
	for i := range a {
		if !a[i].Equals(b[i]) { return false }
	}
	return true
}

File: Rust/ExampleTests/common/inputcontrolvalues_string.rs
#![allow(dead_code, unused_imports, unused_variables)]

#[derive(Debug, Clone, Default)]
pub struct InputControlValuesString {
    pub frame: String,
    pub roll: String,
    pub remaining: String,
}

impl InputControlValuesString {
    pub fn from_vec(v: &[&str]) -> Self {
        Self {
            frame: v.get(0).copied().unwrap_or("").to_string(),
            roll: v.get(1).copied().unwrap_or("").to_string(),
            remaining: v.get(2).copied().unwrap_or("").to_string(),
        }
    }

    /// Builds from the text form, e.g. Money as "25 USD".
    pub fn from_text(text: &str) -> Self {
        let parts = crate::common::tokens::require(text, 3, "InputControlValues");
        Self {
            frame: parts[0].clone(),
            roll: parts[1].clone(),
            remaining: parts[2].clone(),
        }
    }
}

impl std::fmt::Display for InputControlValuesString {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", [crate::common::tokens::token(&self.frame), crate::common::tokens::token(&self.roll), crate::common::tokens::token(&self.remaining)].join(" "))
    }
}

impl PartialEq for InputControlValuesString {
    fn eq(&self, other: &Self) -> bool {
        crate::common::dnc_equal(&self.frame, &other.frame)
            && crate::common::dnc_equal(&self.roll, &other.roll)
            && crate::common::dnc_equal(&self.remaining, &other.remaining)
    }
}
